File: ts/test/tunnels-test/testMultiChannelStream.ts
import { CancellationToken, MultiChannelStream, SshDisconnectReason, SshStream, Stream } from "@microsoft/dev-tunnels-ssh";
import { waitForEvent } from "./promiseUtils";

export class TestMultiChannelStream extends MultiChannelStream {
    constructor(
        public readonly serverStream: Stream, 
    ) {
        super(serverStream);
    }

    public streamsOpened = 0;

    public async openStream(channelType?: string, cancellation?: CancellationToken): Promise<SshStream> {
        const result = super.openStream(channelType, cancellation);
        this.streamsOpened++;
        return result;
    }

    public dropConnection() {
        this.serverStream.dispose();
    }

	public async close(reason?: SshDisconnectReason, message?: string) {
		if (reason !== undefined) {
            await this.session.close(reason, message);
		}

        await super.close();
	}

    public async waitUntilClosed() : Promise<void> {
        if (!this.isClosed) {
            await waitForEvent(this.onClosed);
        }
    }
}

File: cs/test/TunnelsSDK.Test/Mocks/TestMultiChannelStream.cs
using System.Diagnostics;
using Microsoft.DevTunnels.Ssh;

namespace Microsoft.DevTunnels.Test;
public class TestMultiChannelStream : MultiChannelStream
{
    public TestMultiChannelStream(Stream transportStream, TraceSource trace = null)
        : base(transportStream, trace)
    {
    }

    public async Task CloseAsync(SshDisconnectReason reason)
    {
        await Session.CloseAsync(reason);
        Session.Dispose();
        await TransportStream.DisposeAsync();
    }
}
